Blog on Fiat-Shamir Heuristics (#147)
vacp2p/vac.dev: rlog/DRAFT-fiat-shamir.mdx

---
title: 'Vac 101: Transforming an Interactive Protocol to a Noninteractive Argument'
date: 2024-07-01 12:00:00
authors: marvin
published: true
slug: Vac-101-Fiat-Shaimr
categories: research

toc_min_heading_level: 2
toc_max_heading_level: 5
---

> In this post, provide a discussion on a common technique used to convert interactive protocols to their noninteractive variant.

Reviewer: ```suggestion
In this post, we introduce a common technique used to convert interactive protocols to their noninteractive variant.
```

<!--truncate-->

## Introduction

The set of interactive protocols form a class of protocols that consist of communication between two parties: the Prover and the Verifier.
The Prover tries to convince the Verifier of a given claim.
For example, the Prover may want to convince the Verifier that she owns a specific UTXO;
that is, the Prover possesses the ability to spend the UTXO.
In many instances, there is information that the Prover does not wish to reveal to the Verifier.
In our example, it is critical that the Prover does not provide the Verifier with the spending key associated to her UTXO.
In addition to the Prover's claim and secret data, there is additional data, public parameters, that the claimed statement is expressed in terms of.
The public parameters can be thought of as the basis for all similar claims.

> In an interactive protocol, the Prover and the Verifier are in active communication.

Reviewer: Maybe we could include an example of a sigma protocol (for instance). Something basic, just to provide a picture of an interactive protocol.

Author: Moved examples of Schnorr and Chaum-Pedersen up, as well as provided an empty framework diagram (similar to Schnorr)

However, this communication is not practical for many applications.
It is necessary that any party can verify the Prover's claim in decentralized systems.
It is impractical for the Prover to be in active communication with a large number of verifying parties.
Instead, it is desirable for the Prover to generate a proof on their own that can convince any party.
To achieve this, it is necessary for the Prover to generate the Verifier's messages in such a way that the Prover cannot manipulate the Verifier's messages for her benefit.
The [Fiat-Shamir heuristic](https://dl.acm.org/doi/10.5555/36664.36676) is used for this purpose.

## Hash Functions
Cryptographic hash functions serve as the backbone to the Fiat-Shamir heuristic.
A hash function, $\mathsf{Hash}$, is a special function that takes in an arbitrary binary string and outputs a binary string of a predetermined fixed length.
Specifically,
$\mathsf{Hash} : \{0,1\}^* \rightarrow \{0,1\}^n$.

A cryptographic hash function satisfies the following properties:
- **Succinct**: The hash function should be easy to compute; the hash $\mathsf{Hash}({\bf{b}})$ can be efficiently computed for any binary string ${\bf{b}}$.

> - **Preimage Resistance**: It should be impossible to work backwards given the output of a hash function. Let ${\bf{y}}$ be a binary string of length $n$.

Reviewer: ```suggestion
- **Preimage Resistance**: It should be computationally infeasible to work backwards given the output of a hash function. Let ${\bf{y}}$ be a binary string of length $n$.
```

    It should be 'impossible' to find some binary string ${\bf{x}}$ so that ${\bf{y}} = \mathsf{Hash}({\bf{x}})$. 

> - **Collision Resistance**: It should be difficult to find two strings that hash to the same value. It should be 'impossible' to find two binary strings ${\bf{x}_1}$ and ${\bf{x}_2}$ so that $\mathsf{Hash}({\bf{x}_1}) = \mathsf{Hash}({\bf{x}_2}).$

Reviewer: ```suggestion
- **Collision Resistance**: It should be difficult to find two strings that hash to the same value. It should be computationally infeasible to find two binary strings ${\bf{x}_1}$ and ${\bf{x}_2}$ so that $\mathsf{Hash}({\bf{x}_1}) = \mathsf{Hash}({\bf{x}_2}).$
```

> - **Collision Resistance**: It should be difficult to find two strings that hash to the same value. It should be 'impossible' to find two binary strings ${\bf{x}_1}$ and ${\bf{x}_2}$ so that $\mathsf{Hash}({\bf{x}_1}) = \mathsf{Hash}({\bf{x}_2}).$

Reviewer: You could also include a couple of lines introducing computational infeasibility. Also the difference between hash functions and one-way functions.

Author: Added

> Commonly used hash functions include SHA256, Keccak, and Poseidon.

Reviewer: Since these are _the three musketeers_ I would add references to them.

If $m_0, \dots, m_t$ denote the public values in the protocol thus far,
then the Verifier's message is computed as $m_{t+1} := \mathsf{Hash}(m_0|| \cdots || m_t)$.

Since $\mathsf{Hash}$ can be efficiently computed, and the messages $m_0, \dots, m_t$ are public, then any verifying party can compute $m_{t+1}$.
Critically, since $\mathsf{Hash}$ is preimage resistant and collision resistant,
the Prover cannot manipulate her choices of the messages $m_0,\dots, m_t$ to influence the message $m_{t+1}$.
Hence, verifying parties can trust that $m_{t+1}$ is sufficiently random with respect to the preceding messages.

There are two variants of the Fiat-Shamir heuristic: weak and strong.
The weak variant uses all of the publicly traded messages in computing the Verifier's messages but does not include the public parameters.
However, in the strong variant all of the publicly traded messages and public parameters are used to compute the Verifier's messages.
We will provide a discussion on issues that can arise from using the weak Fiat-Shamir heuristic.

### The Schnorr Protocol

> The Schnorr protocol ([9](https://link.springer.com/chapter/10.1007/0-387-34805-0_22)) is usually the first $\Sigma$-protocol that one studies.

Reviewer: If not wrong, this is the first time we mention a sigma protocol. I would suggest introducing them with the example at the beginning of the document (I made a comment with regards to it)

Author: Will do.

Additionally, the Schnorr protocol can be used as an efficient signature scheme.
The Schnorr protocol provides a framework that enables the Prover to convince the Verifier that: for group elements $g$ and $X$, the Prover knows the power to raise $g$ to obtain $X$.

> Specifically, the Prover possesses some integer $x$ so that $X = g^x$.

Reviewer: I'm being punctilious here... but maybe we can add a comment on the additive and the multiplicative notation in cryptography

In particular, Bernhard et al. discuss the pitfalls of the weak Fiat-Shamir heuristic.
Recall that the strong Fiat-Shamir heuristic requires that the public parameters are included in the calculations of the Verifier's messages while the weak version does not.
The inclusion of the public parameters in the hash evaluations fix these public values for the entire protocol.
This means that the Prover cannot retroactively change them.

The issues with the differences in the variants of the Fiat-Shamir heuristics has persisted since Bernhard et al.'s paper. 
In recent years, auditors from [Trail of Bits](https://www.trailofbits.com/) and [OpenZepellin](https://www.openzeppelin.com/) have 
released blogs ([2](https://blog.trailofbits.com/2022/04/13/part-1-coordinated-disclosure-of-vulnerabilities-affecting-girault-bulletproofs-and-plonk/),
 [3](https://blog.trailofbits.com/2022/04/14/the-frozen-heart-vulnerability-in-giraults-proof-of-knowledge/),
 [4](https://blog.trailofbits.com/2022/04/15/the-frozen-heart-vulnerability-in-bulletproofs/), [5](https://blog.trailofbits.com/2022/04/18/the-frozen-heart-vulnerability-in-plonk/), [7](https://blog.openzeppelin.com/the-last-challenge-attack))

>  and papers ([6](https://eprint.iacr.org/2023/691), [8](https://eprint.iacr.org/2024/398)) 

Reviewer 2: Can we add this [paper](https://eprint.iacr.org/2023/1071.pdf) who examines the security of different fiat-shamir on zkSNARKs into the papers.

Author: Added.
